### PR TITLE
Run the test suite

### PR DESCRIPTION
**Pull request by maresb**
Checklist
* [X] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [X] Bumped the build number (if the version is unchanged)
* [X] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [X] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by maresb** [comment]
@conda-forge-admin, please rerender

File: recipe/meta.yaml [hunk 1]
```diff
@@ -33,6 +35,9 @@ requirements:
     - mkl-service
     - cxx-compiler  # [not win]
     - m2w64-toolchain  # [win]
+    - libpython  # [win]
+    - numba  # [not linux]
```
Comment (twiecki) [inline recipe/meta.yaml:39]: numba is not a pymc3 dependency.

File: recipe/meta.yaml [hunk 2]
```diff
@@ -33,6 +35,9 @@ requirements:
     - mkl-service
     - cxx-compiler  # [not win]
     - m2w64-toolchain  # [win]
```
Comment (twiecki) [inline recipe/meta.yaml:37]: I think these dependencies should be in the aesara feedstock, not this one.

File: recipe/meta.yaml [hunk 3]
```diff
@@ -33,6 +35,9 @@ requirements:
     - mkl-service
```
Comment (twiecki) [inline recipe/meta.yaml:35]: This is also optional.

File: recipe/meta.yaml (hunk 1 above)
Comment (maresb) [reply recipe/meta.yaml:39]: I had some reason, I don't remember what though.

My guess is that some test was failing without. But if so, maybe that test was optional?

File: recipe/meta.yaml (hunk 2 above)
Comment (maresb) [reply recipe/meta.yaml:37]: Hmm, maybe it's best to keep them for this version and try to fix in Aesara soon afterwards? After removing this we'd then ≤-pin Aesara to the packages which are fixed?

BTW, I don't see Aesara as a direct dependency here. Is it indirect?

File: recipe/meta.yaml (hunk 3 above)
Comment (maresb) [reply recipe/meta.yaml:35]: Do you mean that numpy and pandas are optional, or something about the pinning is optional?

Comment (twiecki) [reply recipe/meta.yaml:35]: `mkl-service` is not a requirement.

File: recipe/meta.yaml (hunk 2 above)
Comment (twiecki) [reply recipe/meta.yaml:37]: `theano-pymc` was renamed to `aesara`, but pymc3 stable does not yet depend on it. So we might need to wait until 1. aesara has a reliable compile chain dependencies, and 2. we have a pymc3 release that depends on aesara instead of theano-pymc.

File: recipe/meta.yaml (hunk 3 above)
Comment (maresb) [reply recipe/meta.yaml:35]: Ah, sorry, it seems that GitHub glitched and was showing pandas/numpy instead of mkl-service.

**Comment by maresb** [comment]
@twiecki please feel free to commit directly yourself.

**Comment by maresb** [comment]
Some broken pip dependencies:

https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=363667&view=logs&j=9c5ef928-2cd6-52e5-dbe6-9d173a7d951b&t=20c71c51-4b27-578b-485d-06ade2de1d00&l=1271

And under Windows, it's showing a stack overflow. :persevere:

File: recipe/meta.yaml (hunk 1 above)
Comment (michaelosthege) [reply recipe/meta.yaml]: Having numba in the dependencies tricks conda into resolving a compatible of dependencies.
A while back I did really extensive testing, comparing different sets of dependencies, diffing the resulting environments and checking if PyMC3 was importing without warnings...
That's why numba is included in https://github.com/pymc-devs/pymc3/wiki/Installation-Guide-(Windows)

Comment (michaelosthege) [reply recipe/meta.yaml]: I guess they just did a better job at defining the dependencies compared to the authors of Theano/Aesara conda recipes.

**Comment by michaelosthege** [comment]
> And under Windows, it's showing a stack overflow. 😣

The test suite is too large to run in one test job on Windows. It's also reproducible locally that none of these tests fail individually, but give the stack overflow when all of them run in the same `pytest` call.

**Comment by twiecki** [comment]
I think we should figure out what adding `numba` is doing and solving the underlying issue, because it really doesn't make any sense to add it here.

**Comment by maresb** [comment]
@twiecki I completely agree, I just want to reach a state where the CI tests are passing.

Now it's saying pytest failed, even though there were no errors and only warnings? Any idea what's going on?

**Comment by maresb** [comment]
I am getting closer, thanks very much @michaelosthege for the suggestion to run the tests individually to avoid the stack overflow under Windows.

Now just Windows is failing with some theano compilation errors:

[3.7](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=365688&view=logs&j=171a126d-c574-5c8c-1269-ff3b989e923d&t=1183ba29-a0b5-5324-8463-2a49ace9e213&l=1360):
> E               Exception: ('Compilation failed (return status=1): C:\\Users\\VSSADM\~1\\AppData\\Local\\Temp\\ccbQJyYb.s: Assembler messages:\r. C:\\Users\\VSSADM\~1\\AppData\\Local\\Temp\\ccbQJyYb.s:262: Error: invalid register for .seh_savexmm\r. ', 'FunctionGraph(Elemwise{ge,no_inplace}(v, <TensorType(float64, (True,))>))')

[3.8](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=365688&view=logs&j=5be07ae1-d8ba-5406-47b6-8e3a3a12f825&t=0bf03e01-0bec-5b85-5316-b1633322e895&l=1357)
> E               Exception: ('Compilation failed (return status=1): C:\\Users\\VSSADM\~1\\AppData\\Local\\Temp\\cckOnRXa.s: Assembler messages:\r. C:\\Users\\VSSADM\~1\\AppData\\Local\\Temp\\cckOnRXa.s:81: Error: invalid register for .seh_savexmm\r. ', 'FunctionGraph(Elemwise{ge,no_inplace}(v, <TensorType(float64, (True,))>))')

[3.9](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=365688&view=logs&j=e5cdccbf-4751-5a24-7406-185c9d30d021&t=66e4a587-a723-53e9-873c-6c9a2c5971ea&l=3450)
> E               Exception: ("Compilation failed (return status=1): C:\\Users\\VSSADM\~1\\AppData\\Local\\Temp\\cctE7GCa.o: In function `run':\r. C:/Users/VssAdministrator/AppData/Local/Theano/compiledir_Windows-10-10.0.14393-SP0-Intel64_Family_6_Model_79_Stepping_1_GenuineIntel-3.9.6-64/tmpv8iwac39/mod.cpp:1612: undefined reference to `dgemm_'\r. C:\\Users\\VSSADM\~1\\AppData\\Local\\Temp\\cctE7GCa.o:mod.cpp:(.rdata$.refptr.dgemm_[.refptr.dgemm_]+0x0): undefined reference to `dgemm_'\r. collect2.exe: error: ld returned 1 exit status\r. ", 'FunctionGraph(BatchedDot(<TensorType(float64, row)>, <TensorType(float64, row)>))')

Do we need to [mess with cxxflags](https://github.com/Theano/Theano/issues/6693#issuecomment-899340454)? Or is this [a BLAS thing](https://github.com/Theano/Theano/issues/6573)? Any ideas @twiecki or @michaelosthege?

**Comment by twiecki** [comment]
This problems was mentioned here: https://github.com/pymc-devs/pymc3/issues/4749 (seems to be a `m2g64-toolchain` issue, can you confirm that it gets installed?), as well as here: https://github.com/Theano/Theano/issues/6693 (solved by a compiler flag).

Perhaps the latter is the safer fix?

**Comment by maresb** [comment]
In the testing environment, `m2w64-toolchain` is indeed installed [here](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=365688&view=logs&j=5be07ae1-d8ba-5406-47b6-8e3a3a12f825&t=0bf03e01-0bec-5b85-5316-b1633322e895&l=932).

I'm not actually sure how one should set compiler flags from conda. Primarily because I don't work with C++ much, so it's quite mysterious to me. And secondly because it seems very risky and complicated. For example, if we mess with the CXXFLAGS environment variable, what if it conflicts with other programs? I'm pretty stuck on a good way forward. Any suggestions?

**Comment by twiecki** [comment]
The compiler flag I think we would set in `aesara` or `pymc3`, we already do this here: https://github.com/pymc-devs/pymc3/blob/main/pymc3/__init__.py#L40

Curious if @michaelosthege has any ideas.

**Comment by michaelosthege** [comment]
> 
> 
> The compiler flag I think we would set in `aesara` or `pymc3`, we already do this here: https://github.com/pymc-devs/pymc3/blob/main/pymc3/__init__.py#L40 Could we test this here?
> 
> Curious if @michaelosthege has any ideas.

I'm a biologist. This is pretty much black magic to me.
My knowledge about compiler flags ends at the point where I know they exist and they do stuff. With the how the compiler does things I guess.

> I  think we should figure out what adding numba is doing and solving the underlying issue, because it really doesn't make any sense to add it here.

Maybe check out this for comparison? https://github.com/conda-forge/numba-feedstock/blob/master/recipe/meta.yaml

**Comment by maresb** [comment]
@conda-forge-admin, please rerender

**Comment by maresb** [comment]
@conda-forge-admin, please rerender

**Comment by maresb** [comment]
@conda-forge-admin, please rerender

**Comment by maresb** [comment]
@conda-forge-admin, please rerender

**Comment by maresb** [comment]
@conda-forge-admin, please rerender

**Comment by maresb** [comment]
I rebased on the new version.

CI seems to be down. I'll probably have to try again tomorrow.

**Comment by twiecki** [comment]
restarted the tests (I think).

**Comment by maresb** [comment]
@conda-forge-admin, please rerender

**Comment by maresb** [comment]
In Python 3.7 there is a [missing DLL](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=367101&view=logs&j=171a126d-c574-5c8c-1269-ff3b989e923d&t=1183ba29-a0b5-5324-8463-2a49ace9e213&l=1305) when loading netcdf4.

In 3.8 and 3.9 there is a nonsensical error about the following [failed assertion](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=367101&view=logs&j=5be07ae1-d8ba-5406-47b6-8e3a3a12f825&t=0bf03e01-0bec-5b85-5316-b1633322e895&l=1268) in `check_logp()`.

```python
        domains = paramdomains.copy()
        domains["value"] = domain
        for pt in product(domains, n_samples=n_samples):
            pt = Point(pt, model=model)
>           assert_almost_equal(
                logp(pt),
                logp_reference(pt),
                decimal=decimal,
                err_msg=str(pt),
            )
E           AssertionError: 
E           Arrays are not almost equal to 6 decimals
E           {'alpha': array(20.), 'beta': array(0.9), 'value': array(100.)}
E           Mismatched elements: 1 / 1 (100%)
E           Max absolute difference: 3.86856262e+25
E           Max relative difference: 4.70326902e-16
E            x: array(-8.225263e+40)
E            y: array(-8.225263e+40)

pymc3\tests\test_distributions.py:611: AssertionError
```

It looks like numpy's definition of being equal to 6 digits means "6 digits after the 1's place" instead of "6 significant digits".

**Comment by maresb** [comment]
@twiecki, sure! I just rebased on master, so hopefully some of the latest changes have fixed some of the issues.

I don't have any time to work on this, but please feel free to commit to this branch.

Also please see the commented-out changes. For instance, we may need to pin the `semver` package. Unfortunately I've fairly completely forgotten where I was at.

**Comment by maresb** [comment]
Windows is segfaulting right away on `tests\test_pickling.py` even though it's running as an individual test. I can move that one to the end and see if the other tests make any more progress.

OSX/3.7 stalled out mysteriously. Maybe it was a glitch.

I'll rebase master and try again...

**Comment by maresb** [comment]
@michaelosthege, on precisely this advice from you, I already split the tests so that [each test file runs individually](https://github.com/conda-forge/pymc3-feedstock/blob/d47ad6a31e7429f909e30c6a8d8d05f364dffc03/recipe/meta.yaml#L60-L108). :wink: Unfortunately it's still segfaulting on Windows, so I was hoping that you might have another clever idea.

**Comment by maresb** [comment]
This is now mostly obsolete, so I'm closing this out.

It could however be a useful reference in the future if we want to run pytest in `pymc-feedstock`.
